ext/uri/php_lexbor.c

```diff
@@ -0,0 +1,611 @@
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_IPV6_TOO_FEW_PIECES:
+				error_str = ZSTR_INIT_LITERAL("Ipv6TooFewPieces", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_IPV4_IN_IPV6_TOO_MANY_PIECES:
+				error_str = ZSTR_INIT_LITERAL("Ipv4InIpv6TooManyPieces", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_IPV4_IN_IPV6_INVALID_CODE_POINT:
+				error_str = ZSTR_INIT_LITERAL("Ipv4InIpv6InvalidCodePoint", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_IPV4_IN_IPV6_OUT_OF_RANGE_PART:
+				error_str = ZSTR_INIT_LITERAL("Ipv4InIpv6OutOfRangePart", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_IPV4_IN_IPV6_TOO_FEW_PARTS:
+				error_str = ZSTR_INIT_LITERAL("Ipv4InIpv6TooFewParts", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_INVALID_URL_UNIT:
+				error_str = ZSTR_INIT_LITERAL("InvalidUrlUnit", false);
+				ZVAL_FALSE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_SPECIAL_SCHEME_MISSING_FOLLOWING_SOLIDUS:
+				error_str = ZSTR_INIT_LITERAL("SpecialSchemeMissingFollowingSolidus", false);
+				ZVAL_FALSE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_MISSING_SCHEME_NON_RELATIVE_URL:
+				error_str = ZSTR_INIT_LITERAL("MissingSchemeNonRelativeUrl", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_INVALID_REVERSE_SOLIDUS:
+				error_str = ZSTR_INIT_LITERAL("InvalidReverseSoldius", false);
+				ZVAL_FALSE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_INVALID_CREDENTIALS:
+				error_str = ZSTR_INIT_LITERAL("InvalidCredentials", false);
+				ZVAL_FALSE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_HOST_MISSING:
+				error_str = ZSTR_INIT_LITERAL("HostMissing", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_PORT_OUT_OF_RANGE:
+				error_str = ZSTR_INIT_LITERAL("PortOutOfRange", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_PORT_INVALID:
+				error_str = ZSTR_INIT_LITERAL("PortInvalid", false);
+				ZVAL_TRUE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_FILE_INVALID_WINDOWS_DRIVE_LETTER:
+				error_str = ZSTR_INIT_LITERAL("FileInvalidWindowsDriveLetter", false);
+				ZVAL_FALSE(&failure);
+				break;
+			case LXB_URL_ERROR_TYPE_FILE_INVALID_WINDOWS_DRIVE_LETTER_HOST:
+				error_str = ZSTR_INIT_LITERAL("FileInvalidWindowsDriveLetterHost", false);
+				ZVAL_FALSE(&failure);
+				break;
+			EMPTY_SWITCH_DEFAULT_CASE()
+		}
+
+		zval error_type;
+		zend_enum_new(&error_type, whatwg_url_validation_error_type_ce, error_str, NULL);
+		zend_update_property(whatwg_url_validation_error_ce, Z_OBJ(error), ZEND_STRL("type"), &error_type);
+		zend_string_release_ex(error_str, false);
+		zval_ptr_dtor(&error_type);
+
+		zend_update_property(whatwg_url_validation_error_ce, Z_OBJ(error), ZEND_STRL("failure"), &failure);
+
+		add_next_index_zval(errors, &error);
```

Comment:
```suggestion
		zval error_type;
		zend_enum_new(&error_type, whatwg_url_validation_error_type_ce, error_str, NULL);
		zend_update_property(whatwg_url_validation_error_ce, Z_OBJ(error), ZEND_STRL("type"), &error_type);
		zval_ptr_dtor(&error_type);

		zend_update_property(whatwg_url_validation_error_ce, Z_OBJ(error), ZEND_STRL("failure"), &failure);

		add_next_index_zval(errors, &error);

		zend_string_release_ex(error_str, false);
```

To me it makes more sense to release the `error_str` at the very bottom of the loop, since it is initialized at the top. It feels odd to do that in-between, which makes it more likely to miss it (leading to use-after-frees after refactoring).

Reply:
I usually try to free stuffs as soon as possible so that I don't have to worry about memleaks when new stop conditions are added before the `free` operation. `Free`ing late is defensive to avoid possible use-after-frees indeed, but on the other hand, it makes it easier to commit memleaks.